File: 94/sol2.py
from collections import deque


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        result = []
        stack = deque()
        node = root
        while True:
            if node:
                stack.append(node)
                node = node.left
            elif len(stack):
                node = stack.pop()
                result.append(node.val)
                node = node.right
            else:
                break
        return result

File: 144/sol2.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func preorderTraversal(root *TreeNode) []int {
    result := []int{}
    if root == nil {
        return result
    }
    stack := make(tnStack, 0)
    stack = stack.Push(root)
    var node *TreeNode
    for len(stack) > 0 {
        stack, node = stack.Pop()
        result = append(result, (node).Val)
        if node.Right != nil {
            stack = stack.Push(node.Right)
        }
        if node.Left != nil {
            stack = stack.Push(node.Left)
        }
    }
    return result
}

type tnStack []*TreeNode

func (s tnStack) Push(x *TreeNode) tnStack {
    return append(s, x)
}

func (s tnStack) Pop() (tnStack, *TreeNode) {
    return s[:len(s) - 1], s[len(s) - 1]
}
